rain_api_core/auth.py

```diff
@@ -0,0 +1,154 @@
+import contextlib
+import dataclasses
+import json
+import urllib.request
+from http.cookies import CookieError, SimpleCookie
+from time import time
+from typing import List, Mapping, Union
+from wsgiref.handlers import format_date_time as format_7231_date
+
+import jwt
+
+from rain_api_core.logging import get_log
+
+log = get_log()
+
+
+@dataclasses.dataclass
+class UserProfile:
+    user_id: str
+    token: str
+    groups: List[str]
+    first_name: str
+    last_name: str
+    email: str
+    iat: int = None
+    exp: int = None
+
+    @staticmethod
+    def from_jwt_payload(payload):
+        return UserProfile(
+            user_id=payload.get('urs-user-id'),
+            token=payload.get('urs-access-token'),
+            groups=payload.get('urs-groups'),
+            first_name=payload.get('first_name'),
+            last_name=payload.get('last_name'),
+            email=payload.get('email'),
+            iat=payload.get('iat'),
+            exp=payload.get('exp')
+        )
+
+
+class JwtManager:
+    def __init__(
+            self,
+            algorithm: str,
+            public_key: str,
+            private_key: str,
+            cookie_name: str,
+            blacklist_endpoint,
+            session_ttl_in_hours: int = 7 * 24
+    ):
+        self.algorithm = algorithm
+        self.public_key = public_key
+        self.private_key = private_key
+        self.cookie_name = cookie_name
+        self.black_list_endpoint = blacklist_endpoint
+        self.session_ttl = session_ttl_in_hours * 60 * 60
+        self.black_list_last_retrieved = 0
+        self.black_list = {}
+
+    def _get_auth_cookie(self, headers: Mapping[str, str]):
+        cookie_string = headers.get('cookie') or headers.get('Cookie') or headers.get('COOKIE')
+        if not cookie_string:
+            return {}
+
+        cookie = SimpleCookie()
+        with contextlib.suppress(CookieError):
+            cookie.load(cookie_string)
+        return cookie.get(self.cookie_name)
+
+    def _decode_jwt(self, token: str):
+        try:
+            return jwt.decode(token.encode(), self.public_key, self.algorithm)
+        except jwt.ExpiredSignatureError:
+            log.info('JWT has expired')
+        except jwt.InvalidSignatureError:
+            log.info('JWT has failed verification')
+        return None
+
+    def _encode_jwt(self, payload: Mapping[str, str]):
+        try:
+            encoded = jwt.encode(payload, self.private_key, self.algorithm)
+        except TypeError:
+            log.error('unable to encode jwt cookie')
+            return ''
+        return encoded
+
+    @staticmethod
+    def _format_date_for_cookie_header(timestamp):
+        return format_7231_date(timestamp)
```

Comment:
This seems unnecessary to me. I'm not a fan of `staticmethod` in the first place, and since it's literally just proxying to function I'd say just call the function directly.

Reply:
yea, this is just a side effect of moving some implementation details, this function used to add the `sessionttl` which required self but doesn't anymore.